### PR TITLE
feat: support for Azure Blob Storage as storage backend

### PR DESCRIPTION
**Pull request by S0KIT**
Implement #113

### CODE REVIEW

File: pkg/storage/azure.go [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+	if namespace == "" {
+		return fmt.Errorf("namespace argument is empty")
+	}
+
+	if name == "" {
+		return fmt.Errorf("name argument is empty")
+	}
+
+	if filename == "" {
+		return fmt.Errorf("name argument is empty")
```
Comment (oliviermichaelis) [inline pkg/storage/azure.go:258]: Should be `filename argument is empty`, but it's also incorrect in `s3.go` and `gcs.go`

File: pkg/storage/azure.go [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+func (s *AzureStorage) UploadMirroredFile(ctx context.Context, provider *core.Provider, fileName string, reader io.Reader) error {
+	prefix := providerStoragePrefix(s.prefix, mirrorProviderType, provider.Hostname, provider.Namespace, provider.Name)
+	key := filepath.Join(prefix, fileName)
+	return s.upload(ctx, key, reader, true)
+}
+
+func (s *AzureStorage) presignedURL(ctx context.Context, key string) (string, error) {
+	info := service.KeyInfo{
+		Start:  to.Ptr(time.Now().UTC().Format(sas.TimeFormat)),
+		Expiry: to.Ptr(time.Now().UTC().Add(4 * time.Hour).Format(sas.TimeFormat)),
```
Comment (oliviermichaelis) [inline pkg/storage/azure.go:328]: Can you explain why this is set to expire in 4h?

**Review by oliviermichaelis — COMMENTED** [review]
First of all a massive thank you for putting this PR together! I've actually started to implement support for Azure blob storage over the last two months, but didn't get very far due to lack of time, so thanks for submitting your PR :)

I reviewed the code and left a couple of questions and nitpicks, everything looks good so far. However, I haven't had the time to test it out yet.

Another question I have is whether you're interested to contribute unit tests for some of the relevant parts of the code? It's definitely not a requirement to get this PR merged :)

**Comment by S0KIT** [comment]
Thanks for your review. I will go through your comments and make the necessary changes. :)

If you need help with testing, just let me know. I'm glad to help with it.

Unit tests are something I need to have a look at, but I think I can contribute some tests. :)

File: README.md [hunk 3]
```diff
@@ -28,9 +28,9 @@ Support for the [Module Registry Protocol](https://www.terraform.io/internals/mo
 * Provider Registry
 * Network mirror for providers
 * Pull-through mirror for providers
-* Support for S3, GCS, and MinIO object storage
+* Support for S3, GCS, MinIO object storage and Azure Blob Storage
```
[the remark this answers is not in the document]
Comment (S0KIT) [reply README.md]: Resolved in be3855c6f6d6cf5146b7cdfeefa40c4f0f381b69

File: go.mod [hunk 4]
```diff
@@ -26,12 +28,20 @@ require (
...
+	github.com/kylelemons/godebug v1.1.0 // indirect
+	github.com/pkg/browser v0.0.0-20240102092130-5ac0b6a4141c // indirect
+)
+
 require (
 	cloud.google.com/go v0.112.0 // indirect
 	cloud.google.com/go/compute v1.24.0 // indirect
 	cloud.google.com/go/compute/metadata v0.2.3 // indirect
+	github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.5.1
 	github.com/agext/levenshtein v1.2.3 // indirect
```
[the remark this answers is not in the document]
Comment (S0KIT) [reply go.mod]: Resolved in 05bcd2816f76af7968d4e71becf3d74a0ce51212

File: pkg/storage/azure.go (hunk 2 above)
Comment (S0KIT) [reply pkg/storage/azure.go:328]: It's just a random value I choose. We can probably set the time to an even lower value like a couple minutes.

File: pkg/storage/azure.go [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/bloberror"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/sas"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/service"
+)
+
+// AzureStorage is a Storage implementation backed by Azure Blob Storage.
+// AzureStorage implements module.Storage and provider.Storage
```
[the remark this answers is not in the document]
Comment (S0KIT) [reply pkg/storage/azure.go]: Resolved in 7296d775f20d679451f2faaf3b080d751a95917f

File: cmd/root.go [hunk 6]
```diff
@@ -76,6 +82,10 @@ func init() {
 GOOGLE_APPLICATION_CREDENTIALS environment variable might be used as alternative.
 For GCS presigned URLs this SA needs the iam.serviceAccountTokenCreator role.`)
 	rootCmd.PersistentFlags().DurationVar(&flagGCSSignedURLExpiry, "storage-gcs-signedurl-expiry", 30*time.Second, "Generate GCS signed URL valid for X seconds. Only meaningful if used in combination with --gcs-signedurl")
+	rootCmd.PersistentFlags().StringVar(&flagAzureStorageAccount, "storage-azure-account", "", "Azure Storage Account to use for the registry")
+	rootCmd.PersistentFlags().StringVar(&flagAzureStorageContainer, "storage-azure-container", "", "Azure Storage Container to use for the registry")
+	rootCmd.PersistentFlags().StringVar(&flagAzureStoragePrefix, "storage-azure-prefix", "", "Azure Storage prefix to use for the registry")
+	rootCmd.PersistentFlags().DurationVar(&flagAzureStorageSignedURLExpiry, "storage-azure-signedurl-expiry", 5*time.Minute, "Generate Azure Storage signed URL valid for X seconds. Only meaningful if used in combination with --azure-signedurl")
```
[the remark this answers is not in the document]
Comment (S0KIT) [reply cmd/root.go]: Resolved in 9b41dbe37734e19ac74b6429823ec2527bfd823c

File: pkg/storage/azure.go (hunk 1 above)
Comment (S0KIT) [reply pkg/storage/azure.go]: Resolved in 4b2455ef1811e096e987d0446b3c1e2981614246

File: pkg/storage/azure.go (hunk 2 above)
Comment (oliviermichaelis) [reply pkg/storage/azure.go:328]: Okay, I need to do some more research to understand the implications :)

**Comment by oliviermichaelis** [comment]
Thanks! Everything looks good so far, I would just like to test everything out myself and if nothing comes up then, it should be ready to be merged.
I'm quite busy this week with another project I have to finish in time, it might take ~1 week until I get to testing it :/

**Comment by oliviermichaelis** [comment]
FYI I did not forget this PR :)

Unfortunately, I had problems authorizing the azure SDK to my Azure account. I spent a couple of hours trying to get it working, but I didn't succeed.
I tried using a Service principal with a secret, username and password, as well as `az login`. I could use the `az` CLI to list the blobs, but the SDK wasn't able to do so.

**Comment by S0KIT** [comment]
Thanks for the feedback. :) 

Did you receive an error? Could be a permission issue on the Storage Account.

**Review by oliviermichaelis — APPROVED** [review]
Hi @S0KIT !
First of all, sorry for the long delay. I just ran `az login` on my machine and I was able to authorize to Azure without problems :thinking:  Last time I tried, I also tried rebooting to start with a clean state, but it didn't help. I think the only difference was that I previously used an account that was re-activated right before I started testing. No idea if that could be problematic.

Anyways, I was able to test now and everything worked flawlessly :tada: 

Thank you for your contribution!

**Pull request merged**